capabilities.c

```diff
@@ -8,36 +8,100 @@
 *  Please see LICENSE file for your rights under this license. */
 
+// Definition of LINUX_CAPABILITY_VERSION_*
+#define FTLDNS
+#include "dnsmasq/dnsmasq.h"
+#undef __USE_XOPEN
 #include "FTL.h"
-#include <sys/capability.h>
+
+static const unsigned int capabilityIDs[]   = { CAP_CHOWN ,  CAP_DAC_OVERRIDE ,  CAP_DAC_READ_SEARCH ,  CAP_FOWNER ,  CAP_FSETID ,  CAP_KILL ,  CAP_SETGID ,  CAP_SETUID ,  CAP_SETPCAP ,  CAP_LINUX_IMMUTABLE ,  CAP_NET_BIND_SERVICE ,  CAP_NET_BROADCAST ,  CAP_NET_ADMIN ,  CAP_NET_RAW ,  CAP_IPC_LOCK ,  CAP_IPC_OWNER ,  CAP_SYS_MODULE ,  CAP_SYS_RAWIO ,  CAP_SYS_CHROOT ,  CAP_SYS_PTRACE ,  CAP_SYS_PACCT ,  CAP_SYS_ADMIN ,  CAP_SYS_BOOT ,  CAP_SYS_NICE ,  CAP_SYS_RESOURCE ,  CAP_SYS_TIME ,  CAP_SYS_TTY_CONFIG ,  CAP_MKNOD ,  CAP_LEASE ,  CAP_AUDIT_WRITE ,  CAP_AUDIT_CONTROL ,  CAP_SETFCAP ,  CAP_MAC_OVERRIDE ,  CAP_MAC_ADMIN ,  CAP_SYSLOG ,  CAP_WAKE_ALARM ,  CAP_BLOCK_SUSPEND ,  CAP_AUDIT_READ };
+static const char*        capabilityNames[] = {"CAP_CHOWN", "CAP_DAC_OVERRIDE", "CAP_DAC_READ_SEARCH", "CAP_FOWNER", "CAP_FSETID", "CAP_KILL", "CAP_SETGID", "CAP_SETUID", "CAP_SETPCAP", "CAP_LINUX_IMMUTABLE", "CAP_NET_BIND_SERVICE", "CAP_NET_BROADCAST", "CAP_NET_ADMIN", "CAP_NET_RAW", "CAP_IPC_LOCK", "CAP_IPC_OWNER", "CAP_SYS_MODULE", "CAP_SYS_RAWIO", "CAP_SYS_CHROOT", "CAP_SYS_PTRACE", "CAP_SYS_PACCT", "CAP_SYS_ADMIN", "CAP_SYS_BOOT", "CAP_SYS_NICE", "CAP_SYS_RESOURCE", "CAP_SYS_TIME", "CAP_SYS_TTY_CONFIG", "CAP_MKNOD", "CAP_LEASE", "CAP_AUDIT_WRITE", "CAP_AUDIT_CONTROL", "CAP_SETFCAP", "CAP_MAC_OVERRIDE", "CAP_MAC_ADMIN", "CAP_SYSLOG", "CAP_WAKE_ALARM", "CAP_BLOCK_SUSPEND", "CAP_AUDIT_READ"};
+static const unsigned int numCaps = sizeof(capabilityIDs) / sizeof(const int);
 
 bool check_capabilities()
 {
-	if(!cap_get_bound(CAP_NET_ADMIN))
+	// First assume header version 1
+	int capsize = 1; // VFS_CAP_U32_1
+	cap_user_data_t data = NULL;
+	cap_user_header_t hdr = calloc(sizeof(*hdr), capsize);
+
+	// Determine capabilities version used by the current kernel
+	memset(hdr, 0, sizeof(*hdr));
+	capget(hdr, NULL);
+
+	// Check version
+	if (hdr->version != LINUX_CAPABILITY_VERSION_1)
+	{
+		// If unknown version, use largest supported version (3)
+		// Version 2 is deprecated according to linux/capability.h
+		if (hdr->version != LINUX_CAPABILITY_VERSION_2)
+		{
+			hdr->version = LINUX_CAPABILITY_VERSION_3;
+			capsize = 2; // VFS_CAP_U32_3
+		}
+		else
+		{
+			// Use version 2
+			capsize = 2; // VFS_CAP_U32_2
+		}
+	}
+
+	// Get current capabilities
+	data = calloc(sizeof(*data), capsize);
+	capget(hdr, data);
+
+	if(config.debug & DEBUG_CAPS)
+	{
+		logg("***************************************");
+		logg("* Linux capability debugging enabled  *");
+		for(unsigned int i = 0u; i < numCaps; i++)
+		{
+			const unsigned int capid = capabilityIDs[i];
+
+			// Check if capability is valid for the current kernel
+			// If not, exit loop early
+			if(!cap_valid(capid))
+				break;
```

Comment:
This should be a continue so that other capabilities could be checked.

Reply:
The macro is defined as
```c++
#define cap_valid(x) ((x) >= 0 && (x) <= CAP_LAST_CAP)
```
in `linux/capability.h`. There can be no valid capabilities after `CAP_LAST_CAP`. I refrained from using `CAP_LAST_CAP` as condition in the loop as this might overrun our `capabilityIDs[]` array on newer kernels.

Reply:
I assume that the compiler will already optimize the code during compile time taking a `min(CAP_LAST_CAP, capabilityIDs)`

Reply:
Ah, ok. I took this line as checking if the kernel supported the capability (not if the capability is known to the kernel). Is it possible that this macro could be customized on some platforms to return false for certain capabilities (ex. if the kernel doesn't support `NET_ADMIN`)?

Reply:
I checked the [POSIX.1e standard ](http://wt.tuxomania.net/topics/1999_06_Posix_1e/download/Posix_1003.1e-990310.pdf) but haven't found any hint that some capabilities might not be available. What I found was (quoting the more readable form that is also in the `man` pages):

>       From kernel 2.5.27 to kernel 2.6.26, capabilities were an optional ker‐
>       nel component, and  could  be  enabled/disabled  via  the  CONFIG_SECU‐
>       RITY_CAPABILITIES kernel configuration option.

Either they are available or not. Nowadays, they will always be available.

>       Since  Linux  3.2,  the  file /proc/sys/kernel/cap_last_cap exposes the
>       numerical value of the highest capability supported by the running ker‐
>       nel; this can be used to determine the highest bit that may be set in a
>       capability set.

This suggests that a maximum number exists, i.e., the implementation of the macro I quoted above is unlikely to be arch-dependent.